hs_core/models.py

```diff
@@ -3225,11 +3223,40 @@ def _is_valid_element(self, element_name):
         allowed_elements = [el.lower() for el in self.get_supported_element_names()]
         return element_name.lower() in allowed_elements
 
-    def _update_repeatable_element(self, element_name, metadata):
-        # make a list of dict that are for a specific element as specified by element_name
+    def update_non_repeatable_element(self, element_name, metadata, property_name=None):
+        for dict_item in metadata:
+            if element_name in dict_item:
+                if property_name is None:
+                    element = getattr(self, element_name, None)
+                else:
+                    element = getattr(self, property_name, None)
+                if element:
+                    self.update_element(element_id=element.id,
+                                        element_model_name=element_name,
+                                        **dict_item[element_name])
+                else:
+                    self.create_element(element_model_name=element_name,
+                                        **dict_item[element_name])
+
+    def update_repeatable_element(self, element_name, metadata, property_name=None):
+        """
+        Creates new metadata elements of type *element_name*. Any existing metadata elements of
+        matching type get deleted first.
+        :param element_name: class name of the metadata element
+        :param metadata: a list of dicts containing data for each of the metadata elements that
+        needs to be created/updated as part of bulk update
+        :param property_name: (Optional) the property name used in this instance of CoreMetaData
+        (or its sub class) to access all the objects of type *element_type*
+        :return:
+        """
+
         element_list = [element_dict for element_dict in metadata if element_name in element_dict]
         if len(element_list) > 0:
-            elements = getattr(self, element_name + 's')
+            if property_name is None:
+                elements = getattr(self, element_name + 's')
```

Comment:
@hyi The ```element_name``` refers to the metadata element class name. For example, element_name ='creator' refers to class ```Creator```.  Normally the repeatable elements inside ```CoreMetaData``` or subclasses of it use a plural form of attributes (.eg., creators). But in some resource types repeatable elements don't use a plural form for the attribute to refer to a collection of the elements.  By using the optional ```property_name``` I am passing the actual attribute name as used in any of the sub classes of ```CoreMetaData``` so that existing elements can be deleted before creating new ones.